volume-providers/vpc/provider/detach_volume.go

```diff
@@ -0,0 +1,72 @@
+/*******************************************************************************
+ * IBM Confidential
+ * OCO Source Materials
+ * IBM Cloud Container Service, 5737-D43
+ * (C) Copyright IBM Corp. 2018, 2019 All Rights Reserved.
+ * The source code for this program is not  published or otherwise divested of
+ * its trade secrets, irrespective of what has been deposited with
+ * the U.S. Copyright Office.
+ ******************************************************************************/
+
+package provider
+
+import (
+	"github.com/IBM/ibmcloud-storage-volume-lib/lib/provider"
+	userError "github.com/IBM/ibmcloud-storage-volume-lib/volume-providers/vpc/messages"
+	"github.com/IBM/ibmcloud-storage-volume-lib/volume-providers/vpc/vpcclient/models"
+
+	"go.uber.org/zap"
+	"net/http"
+)
+
+// Detach volume based on give volume attachment request
+func (vpcs *VPCSession) Detach(volumeAttachmentTemplate provider.VolumeAttachRequest) (provider.VolumeResponse, error) {
+	vpcs.Logger.Debug("Entry of Detach method...")
+	defer vpcs.Logger.Debug("Exit from Detach method...")
+	var err error
+	vpcs.Logger.Info("Validating basic inputs for detach method...", zap.Reflect("volumeAttachmentTemplate", volumeAttachmentTemplate))
+	err = validateAttachVolumeRequest(volumeAttachmentTemplate)
+	volumeResponse := provider.VolumeResponse{}
+	if err != nil {
+		volumeResponse.Status = provider.FAILURE
+		volumeResponse.Message = err.Error()
+		return volumeResponse, err
+	}
+
+	var response *http.Response
+	// First , check if volume is already attached to given instance
+	vpcs.Logger.Info("Checking if volume is already attached ")
+	currentVolAttachment, _ := vpcs.GetAttachStatus(volumeAttachmentTemplate)
+	if currentVolAttachment.Status == provider.SUCCESS && currentVolAttachment.VPCVolumeAttachment != nil {
```

Comment:
Can we simplify this logic, I would write like this

1- Check volume attach status
2- If its attached 
    then 
     a- detach it
   else
     b- Its not in attached to instance

Reply:
No change is required